### PR TITLE
Support for querying which parameters are used in emitted code

### PR DESCRIPTION
**Pull request by apanteleev**
This is for https://github.com/shader-slang/slang/issues/2213

A new post-emit pass is added, `collectMetadata`, that scans the IR and creates a set of ranges representing the binding indices of resources that were referenced in the IR. Then, a new function `spIsParameterLocationUsed` and its C++ version `ICompileRequest::isParameterLocationUsed` is added to query this information after code generation. 

A new test, `tests/reflection/used-parameters.slang`, is added for this feature. The `-no-codegen` flag was added to most of the reflection tests because the reflection test tool no longer specifies that implicitly, as codegen is needed to test the new used bindings feature.

The `spGetCompileFlags` function was added to have a non-hacky way of supporting this new feature in the reflection test.

### CODE REVIEW

File: source/slang/slang-ir-metadata.cpp [hunk 1]
```diff
@@ -0,0 +1,77 @@
...
+
+    // See if the new range is adjacent to any of the existing ranges, merge with that.
+    for (auto& range : ranges)
+    {
+        if (range.adjacentTo(newRange))
+        {
+            range.mergeWith(newRange);
+            return;
+        }
+    }
```
Comment (tangent-vector) [inline source/slang/slang-ir-metadata.cpp:33]: Does this logic handle the case where two previously non-adjacent ranges become adjacent as a result of the range between them being added? That is, if we add `[0,A)` then `[B,C)` and finally `[A, B)` (with `A < B < C`), will the result have the single range `[0, C)`? It doesn't seem like it from this merge logic.

File: tests/cross-compile/cpp-resource-reflection.slang [hunk 2]
```diff
@@ -1,4 +1,4 @@
-//TEST:CPU_REFLECTION: -profile cs_5_0 -entry computeMain -target cpp 
+//TEST:CPU_REFLECTION: -profile cs_5_0 -entry computeMain -target cpp -no-codegen
```
Comment (tangent-vector) [inline tests/cross-compile/cpp-resource-reflection.slang:1]: I may have missed it in the PR description, but I wonder why some of these compiles need `-no-codegen` now where they were okay without it before.

**Review by tangent-vector — APPROVED** [review]
Looks really good.

File: slang.h [hunk 3]
```diff
@@ -2127,6 +2131,15 @@ extern "C"
...
     SLANG_API unsigned spReflectionParameter_GetBindingSpace(SlangReflectionParameter* parameter);
 
+    SLANG_API SlangResult spIsParameterLocationUsed(
+        SlangCompileRequest* request,
+        SlangInt entryPointIndex,
+        SlangInt targetIndex,
+        SlangParameterCategory category, // is this a `t` register? `s` register?
+        SlangUInt spaceIndex,      // `space` for D3D12, `set` for Vulkan
+        SlangUInt registerIndex,   // `register` for D3D12, `binding` for Vulkan
+        bool& outUsed);
```
[the remark this answers is not in the document]
Comment (apanteleev) [reply slang.h:2141]: D3D also uses this signature type, but mostly when creating objects. Like, `HRESULT ID3D12Device::CreateCommandQueue(pDesc, riid, ppCommandQueue)` - that returns whether the operation was successful, and if not, why not. The actual result is returned through the out parameter. I think it's fine to use the same approach here.

File: slang.h [hunk 4]
```diff
@@ -3872,6 +3890,14 @@ namespace slang
         virtual SLANG_NO_THROW SlangResult SLANG_MCALL getProgramWithEntryPoints(
             slang::IComponentType** outProgram) = 0;
 
+        virtual SLANG_NO_THROW SlangResult SLANG_MCALL isParameterLocationUsed(
+            SlangInt entryPointIndex,
+            SlangInt targetIndex,
+            SlangParameterCategory category,
+            SlangUInt spaceIndex,
+            SlangUInt registerIndex,
+            bool& outUsed) = 0;
```
[the remark this answers is not in the document]
Comment (apanteleev) [reply slang.h:3899]: How would it work though, before linking and optimization (most importantly DCE)?

File: source/slang/slang-ir-metadata.cpp (hunk 1 above)
Comment (apanteleev) [reply source/slang/slang-ir-metadata.cpp:33]: No, there will be two ranges. We can of course run a compaction pass on the ranges after all of them have been collected, but is that necessary?

File: tests/cross-compile/cpp-resource-reflection.slang (hunk 2 above)
Comment (apanteleev) [reply tests/cross-compile/cpp-resource-reflection.slang:1]: Previously, `-no-codegen` was added for all tests implicitly in the reflection test program. Now it's not, as I need to do code generation for the used bindings test. I guess I could drop this option on all/most of the other tests too, and update the expected outputs, which would result in better coverage.

**Pull request merged**